File: backend/src/api/v1/archive/events.py
from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy.orm import Session
from src.database import get_db
from src.services.event_service import (
    get_event_by_id,
    delete_event_by_id,
    create_event,
    update_event,
    get_prices_for_event,
    get_event_price,
)

from src.schemas.event import EventResponse, EventCreate, EventUpdate, PriceResponse
from src.services.auth.permissions import Permissions
from src.api.dependencies import RequirePermissions
from src.models.user import User
from src.services.archive import get_base_url

router = APIRouter()


@router.get("/{event_id}", response_model=EventResponse)
def get_event(event_id: int, request: Request, db: Session = Depends(get_db)):
    base_url = get_base_url(request, 2)

    try:
        event_data = get_event_by_id(db, event_id, base_url)
    except ValueError:
        raise HTTPException(status_code=404, detail="Event not found")

    return event_data


@router.delete("/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_event(
    event_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(RequirePermissions([Permissions.ARCHIVE_DELETE])),
):
    success = delete_event_by_id(db, event_id)

    if not success:
        raise HTTPException(status_code=404, detail="Event not found")


@router.post("/", response_model=EventResponse, status_code=201)
def post_event(
    event_in: EventCreate,
    request: Request,
    db: Session = Depends(get_db),
    _: User = Depends(RequirePermissions([Permissions.ARCHIVE_CREATE])),
):
    base_url = get_base_url(request)

    try:
        return create_event(db, event_in, base_url)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.patch("/{event_id}", response_model=EventResponse)
def patch_event(
    event_id: int,
    update_data: EventUpdate,
    request: Request,
    db: Session = Depends(get_db),
    _: User = Depends(RequirePermissions([Permissions.ARCHIVE_UPDATE])),
):
    base_url = get_base_url(request, 2)
    
    try:
        return update_event(db, event_id, update_data, base_url)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{event_id}/prices", response_model=list[PriceResponse])
def get_event_prices(event_id: int, request: Request, db: Session = Depends(get_db)):
    base_url = get_base_url(request, 3)

    try:
        return get_prices_for_event(db, event_id, base_url)
    except ValueError:
        raise HTTPException(status_code=404, detail="Event not found")


@router.get("/{event_id}/prices/{price_id}", response_model=PriceResponse)
def get_price(
    event_id: int, price_id: int, request: Request, db: Session = Depends(get_db)
):
    base_url = get_base_url(request, 4)

    try:
        return get_event_price(db, event_id, price_id, base_url)
    except ValueError:
        raise HTTPException(status_code=404, detail="Price not found")

File: backend/src/api/v1/archive/events.rs
use axum::{
    extract::{Query, State},
    http::StatusCode,
    Json,
};

use crate::api::dependencies::auth::CurrentUser;
use crate::errors::AppError;
use crate::schemas::event::{EventCreate, EventListQuery, EventResponse};
use crate::services::auth::permissions::{
    check_permission, PERMISSION_ARCHIVE_READ, PERMISSION_ARCHIVE_WRITE,
};
use crate::services::auth::user::get_user_permission_names;
use crate::services::event::{create_event, list_events};
use crate::AppState;

pub async fn list_events_handler(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Query(query): Query<EventListQuery>,
) -> Result<Json<Vec<EventResponse>>, AppError> {
    let perms = get_user_permission_names(&state.db, user.id).await?;
    if !user.super_user {
        check_permission(&perms, PERMISSION_ARCHIVE_READ)?;
    }
    let result = list_events(&state.db, query).await?;
    Ok(Json(result))
}

pub async fn create_event_handler(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Json(data): Json<EventCreate>,
) -> Result<(StatusCode, Json<EventResponse>), AppError> {
    let perms = get_user_permission_names(&state.db, user.id).await?;
    if !user.super_user {
        check_permission(&perms, PERMISSION_ARCHIVE_WRITE)?;
    }
    let result = create_event(&state.db, data).await?;
    Ok((StatusCode::CREATED, Json(result)))
}
